src/Func.h

```diff
@@ -258,6 +262,12 @@ class ScriptFunc : public Func {
     StmtPtr CurrentBody() const { return current_body; }
     int CurrentPriority() const { return current_priority; }
 
+    /**
+     * If this function is an instantiated lambda, returns the primary from
+     * which it was constructed. Otherwise, returns this function itself.
+     */
+    const ScriptFunc* Primary() const { return primary ? primary.get() : this; }
```

Comment:
Hmm, isn't this information already tracked in the AST via `LambaExpr::PrimaryFunc()`? There's also this comment in `LambdaExp::Eval()`:

https://github.com/zeek/zeek/blob/d1e7c8b81ced00378bf430348283ea866c62f2cf/src/Expr.cc#L4250-L4252

It's a bit indirect, but couldn't script_opt leverage that instead of introducing a new API in ScriptFunc - which didn't need to know so far it's a lambda - and for which there's seemingly little value generally?

The following seems to pass the `opt` tests (not sure they cover anything here though) and would allow to skip the `ScriptFunc` changes. Would that work for you?

```
diff --git a/src/script_opt/ProfileFunc.cc b/src/script_opt/ProfileFunc.cc
index 5df70eb6a..fc5abc6cb 100644
--- a/src/script_opt/ProfileFunc.cc
+++ b/src/script_opt/ProfileFunc.cc
@@ -779,6 +779,7 @@ void ProfileFuncs::ComputeBodyHashes(std::vector<FuncInfo>& funcs) {
     for ( auto& l : lambdas ) {
         auto pf = ExprProf(l);
         func_profs[l->PrimaryFunc().get()] = pf;
+        lambda_primaries[l->Name()] = l->PrimaryFunc().get();
         ComputeProfileHash(pf);
     }
 }
@@ -1349,7 +1350,8 @@ bool ProfileFuncs::AssessSideEffects(const SideEffectsOp* se, SideEffectsOp::Acc
 }
 
 std::shared_ptr<SideEffectsOp> ProfileFuncs::GetCallSideEffects(const ScriptFunc* sf) {
-    sf = sf->Primary();
+    if ( lambda_primaries.count(sf->Name()) > 0 )
+        sf = lambda_primaries[sf->Name()];
 
     auto sf_se = func_side_effects.find(sf);
     if ( sf_se != func_side_effects.end() )
diff --git a/src/script_opt/ProfileFunc.h b/src/script_opt/ProfileFunc.h
index af1206623..27c68a215 100644
--- a/src/script_opt/ProfileFunc.h
+++ b/src/script_opt/ProfileFunc.h
@@ -531,6 +531,9 @@ protected:
     // to regular functions, and for that it suffices.
     std::unordered_map<const ScriptFunc*, std::shared_ptr<ProfileFunc>> func_profs;
 
+    // Map lambda names to their primary functions
+    std::unordered_map<std::string, const ScriptFunc*> lambda_primaries;
+
     // Tracks side effects associated with script functions. If we decide in
     // the future to associate richer side-effect information with BiFs then
     // we could expand this to track Func*'s instead.

```

Reply:
Thanks for flagging (and for the diffs!), that's much tidier. I've now pushed this change, assuming I didn't screw up my `git rebase` fu.